torch_xla/experimental/xla_quantized_matmul.py

```diff
@@ -21,83 +21,132 @@ def _check_per_channel_quant_weight_dtype_shapes(input_dim, output_dim, w,
+  w_shape = list(w.shape)
+  assert input_dim % block_size == 0, (
+      f"input_dim should be divisible by block_size, "
+      f"got input_dim: {input_dim}, block_size: {block_size}.")
+  assert w_shape[0] == input_dim / block_size and w_shape[1] == block_size, (
+      f"Weight shape is expected to be [input_dim / block_size, block_size, output_dim], "
+      f"input_dim: {input_dim}, block_size: {block_size}, output_dim: {output_dim}, "
+      f"but got {w_shape}.")
+  assert w_scaler.dim() == 2, (
+      f"weight scaler is expected to be 2D, got {w_scaler.dim()}D Tensor.")
+  assert w_scaler.shape[0] == w_shape[0] and w_scaler.shape[1] == w_shape[-1], (
+      f"weight scaler shape is expect to be [in_channel / block_size, out_channel], "
+      f"got {w_scaler.shape}, weight shape {w_shape}.")
+
+
 @impl(XLA_LIB, "quantized_matmul", "XLA")
 def quantized_matmul_xla(x: torch.Tensor,
                          w: torch.Tensor,
                          scaler: torch.Tensor,
-                         blocksize: int = -1,
+                         block_size: int = -1,
                          int4_weight: bool = False):
   """Quantized Matrix Multiply op on XLA devices.
 
   Args:
       x: torch.Tensor - Activation of Matmul [..., in_channel].
       w: torch.Tensor - Weight Tensor.
          per-channel quant: torch.int8 x [out_channel, in_channel].
+         block_wise quant: torch.int8 x [in_channel / block_size, block_size, out_channel].
       scaler: torch.Tensor - Weight scaler.
          per-channel quant: [out_channel,].
-      blocksize: blocksize for blockwise quantization, -1 for per-channel quantization.
+         blockwise quant: [in_channel / block_size, out_channel].
+      block_size: The blocksize for blockwise quantization, -1 for per-channel quantization.
       int4_weight: if the weights are int4, the int4 weights need to be stored in a int8
                    container (unpacked).
   """
-  assert blocksize == -1, "blockwise quantization is not supported yet."
   if int4_weight:
     # Reinterpret cast the weight to s4 dtype in XLA.
     w = torch_xla._XLAC._xla_cast_int4(w, w.cpu().flatten().numpy().tolist())
-  # Per-channel quant.
-  _check_per_channel_quant_weight_dtype_shapes(x.shape[-1], scaler.shape[0], w,
-                                               scaler)
-  return F.linear(x, w) * scaler
+  if block_size == -1:
+    # Per-channel quant.
+    _check_per_channel_quant_weight_dtype_shapes(x.shape[-1], scaler.shape[0],
+                                                 w, scaler)
+    return F.linear(x, w) * scaler
+  else:
+    # Blockwise quant.
+    _check_blockwise_quant_weight_dtype_shapes(x.shape[-1], w.shape[-1],
+                                               block_size, w, scaler)
+    x = x.reshape(*x.shape[:-1], x.shape[-1] // block_size, block_size)
+    out = torch.einsum('scn,...sc->...sn', w, x)
+    out = torch.einsum('sn,...sn->...n', scaler, out)
```

Comment:
lol I have to admit I don't fully understand the math here but I will assume this is correct given test is passing

Reply:
The internal doc go/ptxla-quantized-ops should explain the math.